python/tvm/hybrid/parser.py

```diff
@@ -0,0 +1,755 @@
+        ast.Add: tir.Add,
+        ast.Sub: tir.Sub,
+        ast.Mult: tir.Mul,
+        ast.Div: tir.Div,
+        ast.FloorDiv: tir.FloorDiv,
+        ast.Mod: tir.FloorMod,
+        ast.BitOr: operator.or_,
+        ast.BitAnd: operator.and_,
+        ast.BitXor: operator.xor,
+        ast.Gt: tir.GT,
+        ast.GtE: tir.GE,
+        ast.Lt: tir.LT,
+        ast.LtE: tir.LE,
+        ast.Eq: tir.EQ,
+        ast.NotEq: tir.NE,
+        ast.And: tir.And,
+        ast.Or: tir.Or,
+    }
+
+    _unaryop_maker = {
+        ast.USub: operator.neg,
+        ast.Invert: operator.invert,
+        ast.Not: tir.Not
+    }
+
+    def __init__(self, src, base_lienno):
+        self.params = None
+        self.buffer_map = None
+        self.dict_attr = None
+        self.scope_emitter = None
+
+        self.src = src.split('\n')
+        self.base_lineno = base_lienno
+        self.current_lineno = 0
+        self.current_col_offset = 0
+        self.meta = None
+
+        self.functions = {}
+
+        self._in_with_func_arg = False
+        self._assign_target = None
+
+    def init_function_parsing_env(self):
+        """Initialize function parsing environment"""
+        self.params = []  # parameter list
+        self.buffer_map = {}  # buffer map
+        self.dict_attr = {}  # dict attr
+        self.scope_emitter = scope_emitter.ScopeEmitter(self)  # scope emitter
+
+    @staticmethod
+    def is_meta(node):
+        """Judge whether an AST node is META"""
+        return isinstance(node, ast.Assign) and len(node.targets) == 1 \
+               and isinstance(node.targets[0], ast.Name) and node.targets[0].id == "__tvm_meta__"
+
+    def init_meta(self, meta_dict):
+        if meta_dict is not None:
+            self.meta = tvm.ir.load_json(json.dumps(meta_dict))
+
+    def visit(self, node):
+        """Override method in ast.NodeVisitor"""
+        old_lineno, old_col_offset = self.current_lineno, self.current_col_offset
+
+        if hasattr(node, "lineno"):
+            self.current_lineno = self.base_lineno + node.lineno - 1
+        if hasattr(node, "col_offset"):
+            self.current_col_offset = node.col_offset
+
+        method = 'visit_' + node.__class__.__name__
+        visitor = getattr(self, method, self.generic_visit)
+        visit_res = visitor(node)
+
+        self.current_lineno, self.current_col_offset = old_lineno, old_col_offset
+
+        return visit_res
+
+    def wrap_line_col(self, message, lineno, col_offset):
+        """Wrap the message with line number and column offset"""
+        src_line = self.src[lineno - self.base_lineno]
+        leading_space = len(src_line) - len(src_line.lstrip(' '))
+        col_offset = col_offset - leading_space
+        src_line = src_line[leading_space:]
+        return "\n  " + src_line + "\n  " + " " * col_offset + "^\n" + "ParserError in line " \
+               + str(lineno) + " : " + message
+
+    def report_error(self, message, lineno=None, col_offset=None):
```

Comment:
Please add a test case of parsing error to make sure report error works properly(e.g. report the right line). Use pytest.raises